### PR TITLE
HotspotSet data structure for rapid checking for mutation overlap

### PR DESCRIPTION
**Pull request by adamabeshouse**
# Checks
- [x] Follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [x] Follows the [Airbnb React/JSX Style guide](https://github.com/airbnb/javascript/tree/master/react).
- [x] Make sure your commit messages end with a Signed-off-by string (this line
  can be automatically added by git if you run the `git-commit` command with
  the `-s` option)

### CODE REVIEW

File: src/shared/lib/HotspotSet.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ *  that is overlapped-with.
+ */
+export default class HotspotSet {
+    private regions:[number,number][];
+    constructor(intervals:[number,number][]) {
+        // intervals: [L:number, U:number][] where L <= U
+        this.regions = createHotspotRegions(intervals);
+    }
+
+    public check = function(x:number, y?:number):boolean {
```
Comment (adamabeshouse) [inline src/shared/lib/HotspotSet.ts]: oops, accidentally left it as `= function`

File: src/shared/lib/HotspotSet.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ *  and it does a binary search to quickly check whether the input range or position
+ *  overlaps with any hotspot regions specified in the constructor.
+ *
+ * The class does not support telling you exactly which input intervals are overlapped-with,
+ *  that information is lost. But it will tell you true or false: whether there was some input interval
+ *  that is overlapped-with.
+ */
+export default class HotspotSet {
+    private regions:[number,number][];
+    constructor(intervals:[number,number][]) {
```
Comment (onursumer) [inline src/shared/lib/HotspotSet.ts]: how about defining a type for interval? Something like this:
```JavaScript
type Interval {
    start: number,
    end: number
}
```

And then define `regions` and `intervals` like this:

```JavaScript
    regions: Interval[]
    intervals: Interval[]
```

**Comment by onursumer** [comment]
@adamabeshouse have you tested `new HotspotSet(...)` in the browser?

I am getting a red screen of death with this error:
`TypeError: hotspotSet.check is not a function`

**Comment by onursumer** [comment]
[Match hotspots properly](https://trello.com/c/OjU5RC58/213-match-hotspots-properly)

**Pull request merged**